### ISSUE TITLE
ConcurrentChain.SetTipNoLock inefficient on reorg due to bug

### ISSUE THREAD

**Issue by Aprogiena** [stratisproject/StratisBitcoinFullNode#314]
1)
```
			foreach (var newBlock in block.EnumerateToGenesis()
				.TakeWhile(c => c != Tip))
```
this is wrong on reorg as Tip refers to old chain block, so there is no block in new chain (as yielded from block.EnumerateToGenesis) that could be equal to Tip, so TakeWhile never ends and the foreach body is executed for ALL blocks, even if the reorg is just 2 blocks.

TakeWhile condition should be `c != fork`

2) 
Also `.AddOrReplace` extensions seem to be too complicated in case the key exist:
```
		public static void AddOrReplace<TKey, TValue>(this IDictionary<TKey, TValue> dico, TKey key, TValue value)
		{
			if(dico.ContainsKey(key))
			{
				dico.Remove(key);
				dico.Add(key, value);
			}
			else
			{
				dico.Add(key, value);
			}
		}
```

I can't see why this is not 

```
		public static void AddOrReplace<TKey, TValue>(this IDictionary<TKey, TValue> dico, TKey key, TValue value)
		{
			if (dico.ContainsKey(key)) dico[key] = value;
			else dico.Add(key, value);
		}
```
But that's NBitcoin, so probably not relevant for Stratis full node ..

**Comment by Aprogiena** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [[Chain] ConcurrentChain.SetTipNoLock inefficient on reorg due to bug](https://trello.com/c/9NVR9b4Q/165-chain-concurrentchainsettipnolock-inefficient-on-reorg-due-to-bug)

**Comment by NicolasDorier** [comment]
mmh this seems about right... should be `c => c != fork` right ?

**Comment by Aprogiena** [comment]
yup

**Comment by NicolasDorier** [comment]
Fixed by d5e1a2509d4fc5f54b91563ce3669994a0b8f649 on NBitcoin
